lib/populator.js

```diff
@@ -42,22 +42,24 @@ Populator.prototype._write = function (chunk, encoding, callback) {
 
 Populator.prototype._processRow = function (row, next) {
   this.emit('processRow');
-  var tasks = [
-    // Populate the map of tracked entity attributes to type
-    Async.each.bind(null, Object.keys(row.attributes), this._getTrackedEntityAttributeType.bind(this)),
-    // Populate the map of data elements to type
-    Async.each.bind(null, Object.keys(row.dataElements), this._getDataElementType.bind(this)),
-    // Track the entity
-    this._addTrackedEntity.bind(this, row.parameters, row.attributes),
+  var tasks = [];
+  // Populate the map of tracked entity attributes to type
+  tasks.push(Async.each.bind(null, Object.keys(row.attributes), this._getTrackedEntityAttributeType.bind(this)));
+  // Populate the map of data elements to type
+  tasks.push(Async.each.bind(null, Object.keys(row.dataElements), this._getDataElementType.bind(this)));
+  // Track the entity
+  tasks.push(this._addTrackedEntity.bind(this, row.parameters, row.attributes));
+  if (this._options.programID && row.parameters.programDate) {
     // Enroll in the program
-    this._enrollInProgram.bind(this, row.parameters),
-    // Add the event
-    this._addEvent.bind(this, row.parameters, row.dataElements)
-  ];
-
-  if (this._options.duplicateThreshold >= 0) {
-    // Check for duplicate events
-    tasks.splice(4, 0, this._checkForDuplicateEvent.bind(this, row.parameters));
+    tasks.push(this._enrollInProgram.bind(this, row.parameters));
+  }
+  if (this._options.stageID && row.parameters.eventDate) {
+    // Track the events
+    tasks.push(this._addEvent.bind(this, row.parameters, row.dataElements));
+    if (this._options.duplicateThreshold >= 0) {
+      // Check for duplicate events
+      tasks.splice(4, 0, this._checkForDuplicateEvent.bind(this, row.parameters));
```

Comment:
Since we've changed the way that the tasks array is built up now we don't need to use splice anymore and can conditionally add the checkForDuplicateEvent task before the addEvent task.